enterprise/backend/test/metabase_enterprise/data_editing/foreign_keys_test.clj

```diff
@@ -0,0 +1,152 @@
+(ns metabase-enterprise.data-editing.foreign-keys-test
+  (:require
+   [clojure.test :refer :all]
+   [metabase-enterprise.data-editing.foreign-keys :as fks]))
+
+(def ^:private metadata
+  {:orders     [{:table :order-items, :fk {:order-id :id}, :pk [:id]}]
+
+   ;; family tree
+
+   :people     [{:table :people, :fk {:father_id :id}, :pk [:id]}
+                {:table :people, :fk {:mother_id :id}, :pk [:id]}]
+
+   ;; project tracker
+
+   :users      [{:table :teams, :fk {:managed-by :id}, :pk [:id]}
+                {:table :users, :fk {:reports-to :id}, :pk [:id]}
+                {:table :programmes, :fk {:guided-by :id}, :pk [:id]}
+                {:table :tasks, :fk {:assigned-to :id}, :pk [:id]}]
+
+   :teams      [{:table :users, :fk {:member-of :id}, :pk [:id]}]
+
+   :programmes [{:table :projects, :fk {:part-of :id}, :pk [:id]}
+                {:table :teams, :fk {:focus :id}, :pk [:id]}]
+
+   :projects   [{:table :tasks, :fk {:project :id} :pk [:id]}]})
+
+(def db
+  {:orders      [{:id 42}
+                 {:id 43}]
+   :order-items [{:id 1337, :order-id 42}
+                 {:id 1338, :order-id 42}
+                 {:id 1339, :order-id 43}]
+
+   ;; family tree
+
+   :people      [{:id 1}
+                 {:id 2}
+                 {:id 3 :father_id 1 :mother_id 2}
+                 {:id 4 :father_id 1 :mother_id 3}
+                 {:id 5 :father_id nil :mother_id 3}]
+
+   ;; project tracker
+
+   :users       [{:id :user/ceo}
+                 {:id :user/cto, :reports-to :user/ceo}
+                 {:id :user/cpo, :reports-to :user/ceo}
+                 {:id :user/em, :reports-to :user/cto, :belongs-to :team/alpha}
+                 {:id :user/pm, :member-of :team/alpha}
+                 {:id :user/alice, :member-of :team/alpha}
+                 {:id :user/bob, :member-of :team/alpha}
+                 {:id :user/clarence, :member-of :team/alpha}]
+
+   :teams       [{:id :team/alpha, :managed-by :user/em, :focus :programme/skunk-works}
+                 {:id :team/bravo, :managed-by :user/cto}]
+
+   :programmes  [{:id :programme/skunk-works, :guided-by :user/cpo}]
+
+   :projects    [{:id :project/gamma, :part-of :programme/skunk-works}]
+
+   :tasks       [{:id :task/foo, :project :project/gamma, :assigned-to :user/alice}]})
+
+(defn matches? [parent fks child]
+  (= (map parent (vals fks))
+     (map child (keys fks))))
+
+(defn lookup-children [relationship parents]
```

Comment:
take the db you operate on then you can drop the dvar and `with-` macro (applies also to `delete-item` but with the db-ref instead).